status/status_ext_test.go

```diff
@@ -80,3 +90,161 @@ func (s) TestErrorIs(t *testing.T) {
+	}} {
+		t.Run(serverType.name, func(t *testing.T) {
+			// Start a simple server that returns the trailer and error it receives from
+			// channels.
+			trailerCh := make(chan metadata.MD, 1)
+			errCh := make(chan error, 1)
+			ss := &stubserver.StubServer{
+				UnaryCallF: func(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+					grpc.SetTrailer(ctx, <-trailerCh)
+					return nil, <-errCh
+				},
+			}
+			if err := serverType.startServerFunc(ss); err != nil {
+				t.Fatalf("Error starting endpoint server: %v", err)
+			}
+			if err := ss.StartClient(); err != nil {
+				t.Fatalf("Error starting endpoint client: %v", err)
+			}
+			defer ss.Stop()
+
+			// Convenience function for making a status including details.
+			detailErr := func(c codes.Code, m string) error {
+				s, err := status.New(c, m).WithDetails(&testpb.SimpleRequest{
+					Payload: &testpb.Payload{Body: []byte("detail msg")},
+				})
+				if err != nil {
+					t.Fatalf("Error adding details: %v", err)
+				}
+				return s.Err()
+			}
+
+			serialize := func(err error) string {
+				buf, _ := proto.Marshal(status.Convert(err).Proto())
+				return string(buf)
+			}
+
+			testCases := []struct {
+				name        string
+				trailerSent metadata.MD
+				errSent     error
+				trailerWant []string
+				errWant     error
+				errContains error
+			}{{
+				name:        "basic without details",
+				trailerSent: metadata.MD{},
+				errSent:     status.Error(codes.Aborted, "test msg"),
+				errWant:     status.Error(codes.Aborted, "test msg"),
+			}, {
+				name:        "basic without details passes through trailers",
+				trailerSent: metadata.MD{"grpc-status-details-bin": []string{"random text"}},
+				errSent:     status.Error(codes.Aborted, "test msg"),
+				trailerWant: []string{"random text"},
+				errWant:     status.Error(codes.Aborted, "test msg"),
+			}, {
+				name:        "basic without details conflicts with manual details",
+				trailerSent: metadata.MD{"grpc-status-details-bin": []string{serialize(status.Error(codes.Canceled, "test msg"))}},
+				errSent:     status.Error(codes.Aborted, "test msg"),
+				trailerWant: []string{serialize(status.Error(codes.Canceled, "test msg"))},
+				errContains: status.Error(codes.Internal, "mismatch"),
+			}, {
+				name:        "basic with details",
+				trailerSent: metadata.MD{},
+				errSent:     detailErr(codes.Aborted, "test msg"),
+				trailerWant: []string{serialize(detailErr(codes.Aborted, "test msg"))},
+				errWant:     detailErr(codes.Aborted, "test msg"),
+			}, {
+				name:        "basic with details discards user's trailers",
+				trailerSent: metadata.MD{"grpc-status-details-bin": []string{"will be ignored"}},
+				errSent:     detailErr(codes.Aborted, "test msg"),
+				trailerWant: []string{serialize(detailErr(codes.Aborted, "test msg"))},
+				errWant:     detailErr(codes.Aborted, "test msg"),
+			}}
+
+			for _, tc := range testCases {
+				t.Run(tc.name, func(t *testing.T) {
```

Comment:
Instead of having nested subtests, which mean that tests share a stubServer, could we instead have a single level of subtests that don't share any state with each other, and be completely hermetic. 

One option would be to have a field in the table which controls whether to spawn a regular server or a handler server, thereby getting rid of the two levels of subtests.

Reply:
nested or not is independent of whether there is a different server for each test case.  I did this to share the cost of starting the server.  If you want there to be new servers for every case, I can do that although I don't think it's super important since the only real danger is if there is one failure then it could cause subsequent failures.

Moved server init inside inner-most `Run`.  This allowed me to remove the channel anyway.

Before (best of 2):
```
$ time go test -count=10 -run Test/StatusDetails ./status
ok  	google.golang.org/grpc/status	0.088s

real	0m0.515s
user	0m0.969s
sys	0m0.526s
```

After (best of 3):
```
$ time go test -count=10 -run Test/StatusDetails ./status
ok  	google.golang.org/grpc/status	0.170s

real	0m0.589s
user	0m1.151s
sys	0m0.554s
```